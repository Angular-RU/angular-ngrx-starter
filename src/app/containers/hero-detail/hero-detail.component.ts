import {
  Component,
  OnInit,
  Input,
  ChangeDetectionStrategy
} from '@angular/core';

import { Store, select } from '@ngrx/store';

import { Observable } from 'rxjs';

import { Hero } from '@appModels/hero';

import * as fromSelectors from '@appStore/selectors';
import * as fromReducer from '@appStore/reducers';

import { Back } from '@appStore/actions/router.actions';
import { UpdateHero } from '@appStore/actions/hero.actions';

@Component({
  selector: 'app-hero-detail',
  templateUrl: './hero-detail.component.html',
  styleUrls: ['./hero-detail.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class HeroDetailComponent implements OnInit {
  hero$: Observable<Hero>;

  constructor(private store: Store<fromReducer.hero.State>) {}

  ngOnInit(): void {
    this.hero$ = this.store.pipe(select(fromSelectors.getHeroById));
  }

  goBack(): void {
    this.store.dispatch(new Back());
  }

  save(hero: Hero, heroName: string): void {
    this.store.dispatch(new UpdateHero({ ...hero, name: heroName }));
  }
}
