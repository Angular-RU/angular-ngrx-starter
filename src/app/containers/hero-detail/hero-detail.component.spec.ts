import { HeroDetailComponent } from './hero-detail.component';
import * as fromReducer from '@appStore/reducers';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Store } from '@ngrx/store';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import {
  instance,
  mock,
  verify,
  anything,
  when,
  spy,
  capture
} from 'ts-mockito';
import { of } from 'rxjs';

describe('HeroDetailComponent', () => {
  let comp: HeroDetailComponent;
  let fixture: ComponentFixture<HeroDetailComponent>;
  let storeMock: Store<fromReducer.hero.State>;

  beforeEach(() => {
    storeMock = mock(Store);

    TestBed.configureTestingModule({
      declarations: [HeroDetailComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [{ provide: Store, useFactory: () => instance(storeMock) }]
    });
    fixture = TestBed.createComponent(HeroDetailComponent);
    comp = fixture.componentInstance;
  });

  it('can load instance', () => {
    expect(comp).toBeTruthy();
  });

  describe('ngOnInit', () => {
    it('makes expected calls', () => {
      comp.ngOnInit();

      verify(storeMock.pipe(anything())).once();
    });
  });

  describe('goBack', () => {
    it('makes expected calls', () => {
      const goBackSpy = spy(fixture.debugElement.injector.get(Store));

      comp.goBack();

      verify(storeMock.dispatch(anything())).once();
      expect(capture(goBackSpy.dispatch).last()).toMatchSnapshot();
    });
  });

  describe('save', () => {
    it('makes expected calls', () => {
      const saveSpy = spy(fixture.debugElement.injector.get(Store));

      comp.save({ id: 1, name: 'test name' }, 'new test name');

      verify(storeMock.dispatch(anything())).once();
      expect(capture(saveSpy.dispatch).last()).toMatchSnapshot();
    });
  });

  describe('html', () => {
    it('show hero', () => {
      when(storeMock.pipe(anything())).thenReturn(
        of({ id: 1, name: 'test hero' })
      );

      fixture.detectChanges();

      expect(fixture.debugElement.nativeElement).toMatchSnapshot();
    });
  });
});
