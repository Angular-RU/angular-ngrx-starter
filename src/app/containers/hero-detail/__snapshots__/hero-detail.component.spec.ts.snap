// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroDetailComponent goBack makes expected calls 1`] = `
Array [
  Back {
    "type": "[Router] Back",
  },
]
`;

exports[`HeroDetailComponent html show hero 1`] = `
<div
  id="root4"
  ng-version="7.1.4"
>
  
  <div>
    <h2>
      TEST HERO Details
    </h2>
    <div>
      <span>
        id: 
      </span>
      1
    </div>
    <div>
      <label>
        name: 
        <input
          placeholder="name"
        />
      </label>
    </div>
    <button>
      go back
    </button>
    <button>
      save
    </button>
  </div>
</div>
`;

exports[`HeroDetailComponent save makes expected calls 1`] = `
Array [
  UpdateHero {
    "payload": Object {
      "id": 1,
      "name": "new test name",
    },
    "type": "[Hero] update hero",
  },
]
`;
